File: C/322_coinChange.c
#define Min(x, y) (x) < (y)? (x) : (y)

int coinChange(int* coins, int coinsSize, int amount){
    int * dp = (int*)malloc((amount + 1) * sizeof(int));
    dp[0] = 0;
    for(int i = 1; i <= amount; ++i){
        dp[i] = amount + 1;
        for(int j = 0; j < coinsSize; ++j){
            if(coins[j] > i) continue;;
            dp[i] = Min(dp[i], 1 + dp[i - coins[j]]);
        }
    }
    return dp[amount] == amount + 1? -1 : dp[amount];
}

File: go/322.coin-change.go
package leetcode

func CoinChange322(coins []int, amount int) int {
	return coinChange(coins, amount)
}

/*
 * @lc app=leetcode id=322 lang=golang
 *
 * [322] Coin Change
 */

// @lc code=start
func coinChange(coins []int, amount int) int {
	// we cannot use math.Inf
	const MAX uint = ^uint(0) - 1
	dp := make([]uint, amount+1)
	for i := range dp {
		dp[i] = MAX
	}

	dp[0] = 0
	for i := uint(1); i <= uint(amount); i++ {
		for _, coin := range coins {
			if uint(coin) > i {
				continue
			}
			// convert to uint to compare
			dp[i] = min(1+dp[i-uint(coin)], dp[i])
		}
	}

	if dp[amount] == MAX {
		return -1
	}

	return int(dp[amount])
}

// @lc code=end
